### ISSUE TITLE
Improve traceback of cancelled tasks / add cancel() msg argument

### ISSUE THREAD

**Issue by 01e27b45-90f2-4c74-9e5e-7e7e54c3d78e** [python/cpython#75216]
BPO | [31033](https://bugs.python.org/issue31033)
--- | :---
Nosy | @carljm, @cjerdonek, @ambv, @socketpair, @1st1, @matrixise, @thehesiod, @miss-islington
PRs | <li>python/cpython#19951</li><li>python/cpython#19979</li><li>python/cpython#20173</li><li>python/cpython#20202</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2020-05-18.05:56:13.267>
created_at = <Date 2017-07-25.13:21:40.773>
labels = ['type-feature', '3.9', 'expert-asyncio']
title = 'Improve traceback of cancelled tasks / add cancel() msg argument'
updated_at = <Date 2020-05-19.02:14:19.517>
user = 'https://github.com/socketpair'
```

bugs.python.org fields:
```python
activity = <Date 2020-05-19.02:14:19.517>
actor = 'lukasz.langa'
assignee = 'none'
closed = True
closed_date = <Date 2020-05-18.05:56:13.267>
closer = 'chris.jerdonek'
components = ['asyncio']
creation = <Date 2017-07-25.13:21:40.773>
creator = 'socketpair'
dependencies = []
files = []
hgrepos = []
issue_num = 31033
keywords = ['patch']
message_count = 23.0
messages = ['299079', '299083', '299084', '299101', '299103', '299107', '299112', '299891', '300152', '314953', '314954', '341967', '368236', '368243', '368405', '368987', '368988', '368993', '369173', '369179', '369180', '369305', '369315']
nosy_count = 8.0
nosy_names = ['carljm', 'chris.jerdonek', 'lukasz.langa', 'socketpair', 'yselivanov', 'matrixise', 'thehesiod', 'miss-islington']
pr_nums = ['19951', '19979', '20173', '20202']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue31033'
versions = ['Python 3.9']
```

</p></details>

**Comment by 01e27b45-90f2-4c74-9e5e-7e7e54c3d78e** [comment]
History:

First, I tried to debug code around asyncio.Task() cancelling. So I wrote:
\=====
try:
   ...
except Exception as e:
   print(e)
\=====

When task was cancelled, an empty string printed. I wondered why. So I change the code to

\====
```py
print(repr(e))
```

====

and it printed 'CancelledError' as expected.

Next, I tried:

\====
```py
print(Exception())
```

====

It prints empty string too!

So I came up to propose API change. I propose to add argument to the .cancel() methods (for Task and for Future). This argument should be passed to the CancelledError constructor. This will greatly improves debugging -- it allows to easily know why Future/Task was cancelled.

Also, this change does not break current code. Argument must be optional.

**Comment by 1st1** [comment]
Well, don't use "print(e)" when you are printing errors in in Python. This is pretty standard behaviour for all Python code, not just asyncio. A lot of code in stdlib raises exceptions, almost none allows to customize them.

**Comment by matrixise** [comment]
you could use gdb or pdb for the debugging.

**Comment by 01e27b45-90f2-4c74-9e5e-7e7e54c3d78e** [comment]
Yes, I agree with you about my weird way of debugging. But anyway, changing API with adding ability to pass actual cause would be welcome.

**Comment by 1st1** [comment]
> Yes, I agree with you about my weird way of debugging. But anyway, changing API with adding ability to pass actual cause would be welcome.

I'm not opposed to the idea, btw. If we do decide to add an argument to 'cancel', we probably should do the same for concurrent.futures.

Another possibility would be to allow cancellation via Future.set_exception:

```py
  task.set_exception(asyncio.CancelledError('message'))
```

**Comment by 01e27b45-90f2-4c74-9e5e-7e7e54c3d78e** [comment]
Hmmm....

```py
task.set_exception(Exception('xxx'))
```


Will it cancel whole chain of depending futures in a RIGHT way ?
Or we must require exception passed here to be subclassed from CancelledError ?

**Comment by 1st1** [comment]
> Will it cancel whole chain of depending futures in a RIGHT way ?

I was thinking about this:

  def Future.set_exception(exc):
      if isinstance(exc, asyncio.CancelledError):
          return self.\_cancel(exc)
      \# here goes old code

  def Future.cancel():
      return self.\_cancel(asyncio.CancelledError())

Although now, that I'm looking at it, I don't like this idea, because setting as exception is a different operation from cancelling a task/future, and we shouldn't mix them.  It's OK to set a CancelledError without cancelling.  Also this would be a backwards incompatible change.

So back to square one -- we can consider passing extra args to .cancel() methods.

**Comment by cjerdonek** [comment]
A couple thoughts on this issue:

First, I think the OP's original issue could perhaps largely be addressed without having to change cancel()'s signature. Namely, simply passing a hard-coded string to CancelledError in the couple spots that CancelledError is raised would cause the exception to display:

https://github.com/python/cpython/blob/733d0f63c562090a2b840859b96028d6ec0a4803/Lib/asyncio/futures.py#L153
https://github.com/python/cpython/blob/733d0f63c562090a2b840859b96028d6ec0a4803/Lib/asyncio/futures.py#L173

The "raise CancelledError" could be changed to "raise CancelledError('future is cancelled')", a bit like how InvalidStateError is handled a couple lines later:

```py
    if self._state == _CANCELLED:
        raise CancelledError
    if self._state != _FINISHED:
        raise InvalidStateError('Result is not ready.')
```


Second, in terms of making cancellations easier to debug, is it a deliberate design decision that the CancelledError traceback "swallows" / doesn't show the point at which the coroutine was cancelled?

For example, running the following code--

```py
    async def run():
        await asyncio.sleep(1000000)

    loop = asyncio.new_event_loop()
    task = asyncio.ensure_future(run(), loop=loop)
    loop.call_later(2, task.cancel)
    loop.run_until_complete(task)
```


Results in the following output:

```pytb
    Traceback (most recent call last):
      File "test-cancel.py", line 46, in <module>
        loop.run_until_complete(task)
      File "/Users/.../python3.6/asyncio/base_events.py", line 466,
          in run_until_complete
        return future.result()
    concurrent.futures._base.CancelledError
```


In particular, it doesn't show that the task was waiting on asyncio.sleep(1000000) when the task was cancelled. It would be very useful to see full tracebacks in these cases. (Sorry in advance if this second point is off-topic for this issue.)

**Comment by 01e27b45-90f2-4c74-9e5e-7e7e54c3d78e** [comment]
1. Yes, specifying argument to cancel in `raise CancelledError` would be perfect
2. specifying argument to `.cancel()` is still actual.
3. Yes, important part of exceptions should not be swallowed (please open separate issue). Also, it will be nice to show which code calls `.cancel()`, not only `sleep()` you menitioned.
